File: bitset.py
class Bitset:
    def __init__(self, N):
        self.N = N
        self.bits = [0] * ((N + 31) // 32)

    def get_index(self, position):
        return position // 32

    def get_position(self, position):
        return position % 32

    def set(self, position, value=True):
        if position < self.N:
            idx = self.get_index(position)
            bit_idx = self.get_position(position)

            if value:
                self.bits[idx] |= (1 << bit_idx)
            else:
                self.bits[idx] &= ~(1 << bit_idx)

    def get(self, position):
        if position < self.N:
            idx = self.get_index(position)
            bit_idx = self.get_position(position)
            return (self.bits[idx] >> bit_idx) & 1
        return False

    def toggle(self, position):
        if position < self.N:
            idx = self.get_index(position)
            bit_idx = self.get_position(position)
            self.bits[idx] ^= (1 << bit_idx)

    def reset(self):
        self.bits = [0] * len(self.bits)

    def print(self):
        for i in range(self.N):
            print(self.get(i), end="")
        print()

    def size(self):
        return self.N


# Test
bitset = Bitset(5)

bitset.set(2)
bitset.set(4)
bitset.set(1, False)

bitset.print()

bitset.toggle(4)
bitset.print()

bitset.reset()
bitset.print()

File: bitset.c
#include <stdio.h>
#include <stdlib.h>

#define SIZE 5

typedef struct {
    unsigned int *bits;
} Bitset;

size_t getIndex(size_t position) {
    return position / 32;
}

size_t getPosition(size_t position) {
    return position % 32;
}

void init(Bitset *bitset) {
    bitset->bits = (unsigned int *)calloc((SIZE + 31) / 32, sizeof(unsigned int));
}

void set(Bitset *bitset, size_t position, int value) {
    if (position < SIZE) {
        size_t idx = getIndex(position);
        size_t bit_idx = getPosition(position);

        if (value) {
            bitset->bits[idx] |= (1 << bit_idx);
        } else {
            bitset->bits[idx] &= ~(1 << bit_idx);
        }
    }
}

int get(const Bitset *bitset, size_t position) {
    if (position < SIZE) {
        size_t idx = getIndex(position);
        size_t bit_idx = getPosition(position);
        return (bitset->bits[idx] >> bit_idx) & 1;
    }
    return 0;
}

void toggle(Bitset *bitset, size_t position) {
    if (position < SIZE) {
        size_t idx = getIndex(position);
        size_t bit_idx = getPosition(position);
        bitset->bits[idx] ^= (1 << bit_idx);
    }
}

void reset(Bitset *bitset) {
    for (size_t i = 0; i < (SIZE + 31) / 32; ++i) {
        bitset->bits[i] = 0;
    }
}

void print(const Bitset *bitset) {
    for (size_t i = 0; i < SIZE; ++i) {
        printf("%d", get(bitset, i));
    }
    printf("\n");
}

int main() {
    Bitset bitset;
    init(&bitset);

    set(&bitset, 2, 1);
    set(&bitset, 4, 1);
    set(&bitset, 1, 0);

    print(&bitset);

    toggle(&bitset, 4);
    print(&bitset);

    reset(&bitset);
    print(&bitset);

    free(bitset.bits);
    return 0;
}
